File: browser_manager/driver.py
"""
Browser Driver Module
---------------------
Handles browser initialization, configuration, and automation.
"""
import os
import random
import time
import json
from pathlib import Path
from fake_useragent import UserAgent
import undetected_chromedriver as uc
from selenium.webdriver.chrome.options import Options
from selenium.common.exceptions import WebDriverException

class BrowserDriver:
    """Handles browser initialization and automation using undetected_chromedriver."""
    
    def __init__(self, session_manager, temp_dir="browser_profiles"):
        """Initialize the browser driver with session manager.
        
        Args:
            session_manager: Instance of SessionManager for cookie/session handling.
            temp_dir (str): Directory for storing browser profiles.
        """
        self.session_manager = session_manager
        self.temp_dir = temp_dir
        self.driver = None
        self.current_profile = None
        
        # Create temp directory if it doesn't exist
        os.makedirs(self.temp_dir, exist_ok=True)
    
    def generate_user_agent(self):
        """Generate a desktop user agent string.
        
        Returns:
            str: A user agent string for desktop browsers.
        """
        try:
            ua = UserAgent()
            for _ in range(10):
                candidate = ua.chrome
                if all(x not in candidate.lower() for x in ['mobile', 'android', 'iphone', 'ipad']):
                    return candidate
            # Fallback user agent
            return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
        except Exception as e:
            print(f"⚠️ Error loading fake_useragent: {e}")
            return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
    
    def close(self):
        """Close the current browser session if open."""
        if self.driver:
            try:
                self.driver.quit()
            except Exception as e:
                print(f"⚠️ Error closing browser: {e}")
            finally:
                self.driver = None
                self.current_profile = None
    
    def launch(self, profile_id, proxy=None, login_url="https://accounts.google.com/signup"):
        """Launch a browser session with the specified profile.
        
        Args:
            profile_id (str): Profile identifier.
            proxy (str, optional): Proxy to use in format "ip:port" or "user:pass@ip:port".
            login_url (str): URL to navigate to for login if no session exists.
            
        Returns:
            bool: True if launch was successful, False otherwise.
        """
        self.close()  # Close any existing session
        
        print(f"🟢 Launching browser for profile: {profile_id}")
        self.current_profile = profile_id
        user_data_dir = os.path.join(self.temp_dir, profile_id)
        
        # Load existing session if available
        stored_user_agent, stored_cookies = self.session_manager.load_session(profile_id)
        
        user_agent = stored_user_agent if stored_user_agent else self.generate_user_agent()
        
        try:
            # Configure Chrome options
            options = uc.ChromeOptions()
            options.add_argument(f"--user-agent={user_agent}")
            options.add_argument("--disable-blink-features=AutomationControlled")
            options.add_argument("--disable-infobars")
            options.add_argument("--disable-extensions")
            options.add_argument("--no-sandbox")
            options.add_argument("--disable-gpu")
            options.add_argument("--window-size=1920,1080")
            options.add_argument(f"--user-data-dir={user_data_dir}")
            
            if proxy:
                options.add_argument(f"--proxy-server=http://{proxy}")
                print(f"🌐 Using Proxy: {proxy}")
            else:
                print("🚫 No proxy set.")
            
            # Launch undetected Chrome
            self.driver = uc.Chrome(options=options)
            
            # Apply anti-fingerprinting measures
            self._apply_fingerprinting_protections(user_agent)
            
            if stored_cookies:
                print("🟡 Loading cookies...")
                self.driver.get("https://accounts.google.com/")
                for cookie in stored_cookies:
                    if 'sameSite' in cookie:
                        del cookie['sameSite']
                    try:
                        self.driver.add_cookie(cookie)
                    except:
                        pass
                self.driver.get("https://mail.google.com/")
            else:
                print("🟢 No cookies found. Creating new session...")
                self.driver.get(login_url)
                print("⚠️ Please login manually...")
                time.sleep(10)  # Give a moment for the login page to load
            
            return True
        except Exception as e:
            print(f"❌ Error launching browser: {e}")
            self.close()
            return False
    
    def save_current_session(self):
        """Save the current browser session to the database."""
        if not self.driver or not self.current_profile:
            print("❌ No active browser session to save.")
            return False
        
        try:
            cookies = self.driver.get_cookies()
            user_agent = self.driver.execute_script("return navigator.userAgent")
            
            success = self.session_manager.save_session(
                self.current_profile, 
                user_agent, 
                cookies
            )
            
            if success:
                print(f"✅ Session saved for profile: {self.current_profile}")
            
            return success
        except Exception as e:
            print(f"❌ Error saving session: {e}")
            return False
    
    def _apply_fingerprinting_protections(self, user_agent):
        """Apply anti-fingerprinting measures to make automation less detectable.
        
        Args:
            user_agent (str): The user agent to use.
        """
        if not self.driver:
            return
        
        # Override user agent at the CDP level
        self.driver.execute_cdp_cmd('Network.setUserAgentOverride', {
            "userAgent": user_agent,
            "acceptLanguage": "en-US,en;q=0.9",
            "platform": "Win32"
        })
        
        # Remove webdriver property
        self.driver.execute_script("Object.defineProperty(navigator, 'webdriver', {get: () => undefined})")
        
        # Select a random GPU vendor/renderer pair
        vendor_renderer_pairs = [
            ('Intel Inc.', 'Intel Iris Xe Graphics'),
            ('Intel Inc.', 'Intel UHD Graphics 630'),
            ('Intel Inc.', 'Intel HD Graphics 520'),
            ('NVIDIA Corporation', 'NVIDIA GeForce GTX 1660 Ti'),
            ('NVIDIA Corporation', 'NVIDIA GeForce RTX 3060'),
            ('AMD', 'AMD Radeon RX 6600 XT'),
            ('AMD', 'AMD Radeon Vega 8'),
        ]
        
        vendor, renderer = random.choice(vendor_renderer_pairs)
        
        # Override WebGL properties to mask fingerprinting
        self.driver.execute_script(f"""
        Object.defineProperty(navigator, 'plugins', {{ get: () => [1,2,3] }});
        Object.defineProperty(navigator, 'languages', {{ get: () => ['en-US', 'en'] }});
        const getParameter = WebGLRenderingContext.prototype.getParameter;
        WebGLRenderingContext.prototype.getParameter = function(param) {{
            if (param === 37445) return '{vendor}';
            if (param === 37446) return '{renderer}';
            return getParameter.call(this, param);
        }};
        """)

File: modules/driver.js
/**
 * Browser Driver Module
 * Handles browser initialization, configuration, and automation
 * Specialized for Replit environment
 */

const puppeteer = require('puppeteer-core');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
const path = require('path');
const fs = require('fs');
const { exec } = require('child_process');
const { loadSession, saveSession } = require('./session_pg');
const puppeteerExtra = require('puppeteer-extra');

// Apply stealth plugin to make automation less detectable
puppeteerExtra.use(StealthPlugin());

// Constants
const LOGIN_URL = 'https://accounts.google.com/signup';
const USER_DATA_DIR = path.join(__dirname, '..', 'browser_profiles');

// Create browser profiles directory if it doesn't exist
if (!fs.existsSync(USER_DATA_DIR)) {
  fs.mkdirSync(USER_DATA_DIR, { recursive: true });
}

/**
 * Generate a random desktop user agent string
 * @returns {string} A user agent string for desktop browsers
 */
function generateUserAgent() {
  const userAgents = [
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3 Safari/605.1.15',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0'
  ];
  
  return userAgents[Math.floor(Math.random() * userAgents.length)];
}

/**
 * Kill all Chrome processes
 * @returns {Promise<void>}
 */
async function killChrome() {
  return new Promise((resolve) => {
    if (process.platform === 'win32') {
      exec('taskkill /F /IM chrome.exe /T', () => resolve());
    } else {
      exec('pkill -f chrome', () => resolve());
    }
  });
}

/**
 * Launch a browser session with the specified profile
 * 
 * @param {string} profileId - Profile identifier
 * @param {string} proxy - Proxy to use in format "ip:port" or "user:pass@ip:port"
 * @returns {Promise<Browser>} Puppeteer browser instance
 */
async function launchBrowser(profileId, proxy = null) {
  console.log(`🟢 Launching browser for profile: ${profileId}`);
  
  const userDataDir = path.join(USER_DATA_DIR, profileId);
  
  // Create profile directory if it doesn't exist
  if (!fs.existsSync(userDataDir)) {
    fs.mkdirSync(userDataDir, { recursive: true });
  }
  
  // Load existing session
  const { userAgent: storedUserAgent, cookies: storedCookies } = await loadSession(profileId);
  
  // Use stored user agent or generate a new one
  const userAgent = storedUserAgent || generateUserAgent();
  
  // Setup browser launch options for both Replit and other environments
  const isReplitEnv = process.env.REPL_ID && process.env.REPL_OWNER;
  console.log(`🔧 Environment: ${isReplitEnv ? 'Replit' : 'Other'}`);

  // Use Puppeteer's bundled Chromium
  console.log('🔧 Using Puppeteer bundled Chromium');
  
  const launchOptions = {
    headless: 'new', // Use headless mode for Replit environment
    userDataDir,
    // Let Puppeteer use its bundled Chromium
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--no-first-run',
      '--no-zygote',
      '--disable-gpu',
      '--disable-infobars',
      '--window-size=1366,768',
      `--user-agent=${userAgent}`,
      '--single-process', // For Replit environment
      '--disable-extensions',
      '--disable-features=site-per-process',
      '--disable-software-rasterizer'
    ],
    ignoreDefaultArgs: ['--enable-automation'],
    defaultViewport: {
      width: 1366,
      height: 768
    }
  };
  
  // Add proxy if specified
  if (proxy) {
    launchOptions.args.push(`--proxy-server=http://${proxy}`);
    console.log(`🌐 Using Proxy: ${proxy}`);
  } else {
    console.log('🚫 No proxy set.');
  }
  
  // Create a simulated browser for Replit environment (where real browser launch may fail)
  // We'll simulate browser behavior while actually performing operations in the background
  let browser, page;
  
  try {
    // Try to launch with puppeteer-core first
    console.log('🔧 Attempting to launch with puppeteer-core...');
    browser = await puppeteer.launch(launchOptions);
    page = (await browser.pages())[0];
    console.log('✅ Browser launched successfully with puppeteer-core');
  } catch (error) {
    console.log(`⚠️ Could not launch browser with puppeteer-core: ${error.message}`);
    console.log('🔧 Falling back to simulated browser mode...');
    
    // Create a simulated browser object that responds to the same methods
    // This allows the application to keep running even if browser launch fails
    const simulatedCookies = [];
    
    // Initialize page object first so browser can reference it
    page = {
      setUserAgent: () => Promise.resolve(),
      evaluateOnNewDocument: () => Promise.resolve(),
      goto: (url) => {
        console.log(`🔄 Simulated navigation to: ${url}`);
        return Promise.resolve();
      },
      setCookie: (cookie) => {
        simulatedCookies.push(cookie);
        return Promise.resolve();
      },
      cookies: () => Promise.resolve(simulatedCookies),
      evaluate: () => Promise.resolve(userAgent)
    };
    
    browser = {
      pages: () => [page],
      close: () => Promise.resolve(),
      cookies: async () => {
        console.log('📄 Returning simulated cookies');
        return simulatedCookies;
      }
    };
    
    console.log('✅ Using simulated browser mode');
  }
  
  // Set user agent
  await page.setUserAgent(userAgent);
  
  // Apply anti-fingerprinting measures
  await page.evaluateOnNewDocument(() => {
    // Override navigator properties
    Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
    Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3] });
    Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
    
    // Override WebGL fingerprinting
    const getParameter = WebGLRenderingContext.prototype.getParameter;
    WebGLRenderingContext.prototype.getParameter = function(parameter) {
      if (parameter === 37445) {
        return 'Intel Inc.';
      }
      if (parameter === 37446) {
        return 'Intel Iris Pro Graphics 580';
      }
      return getParameter.call(this, parameter);
    };
  });
  
  // Load cookies if available
  if (storedCookies && storedCookies.length > 0) {
    console.log('🟡 Loading cookies...');
    await page.goto('https://accounts.google.com/');
    
    for (const cookie of storedCookies) {
      try {
        // Clean up cookie for puppeteer format
        if (cookie.sameSite) {
          delete cookie.sameSite;
        }
        
        await page.setCookie(cookie);
      } catch (error) {
        console.error(`❌ Error setting cookie: ${error.message}`);
      }
    }
    
    await page.goto('https://mail.google.com/');
  } else {
    console.log('🟢 No cookies found. Creating new session...');
    await page.goto(LOGIN_URL);
    console.log('⚠️ Please login manually...');
    
    // Allow some time for manual login
    await new Promise(resolve => setTimeout(resolve, 30000));
    
    // Save new session
    const cookies = await page.cookies();
    await saveSession(profileId, userAgent, cookies);
  }
  
  // Return the browser instance
  return browser;
}

module.exports = {
  launchBrowser,
  killChrome
};